### ISSUE TITLE
Ordinals - deeply nested ordinals in the Veblen hierarchy have exponentially slow comparison

### ISSUE THREAD

**Issue by komiamiko** [komiamiko/doodads#4]
This performance issue was identified in [commit da84b916c501d3e980fc611ba89a851524ec0d7b](https://github.com/komiamiko/doodads/commit/da84b916c501d3e980fc611ba89a851524ec0d7b#diff-76edf891dd6accae99312d0f55e35024).

# How to reproduce the issue

Create 2 very similar deeply nested ordinals, ex. *ω^ω^ω^...^(ε_0+1)* and *ω^ω^ω^...^(ε_0+2)*. A height of 20 is sufficient. To quickly create such an ordinal, we can make use of the fundamental sequences, ex.

```py
A = veblen(1,1)[20]
B = veblen(1,1)[21]
```

Actually, this is exactly the method that was used where the issue was spotted.

Then, just perform a comparison on them, ex.

```py
A < B
```

The comparison takes time exponential in the height of the tree.

# Why this should not be possible

Comparison of ordinals should have time bounded linearly in the "complexity" of the ordinals. More formally, for a reasonable notation like the Veblen normal form, and certainly the internal representation, the total number of terms gives the complexity.

Instead, somehow, the runtime grows exponentially.

# The current suspect

While not confirmed or solved yet, there is currently a suspect for the kind of issue.

Basically, at a point in the evaluation tree, children are evaluated twice (or more). This would lead to a runtime of 2^N or worse in the depth of the tree.

Example of how this might come up:

```py
def tuple_cmp(A, B):
  for a,b in zip(A,B):
    if a < b:return -1
    if a > b:return 1
    # here, a and b may be compared 2 times
  return len(A) - len(B)
```

The current 3-way Veblen hierarchy comparison does not mitigate the problem. A larger refactoring of the library to back its "nice" comparison functions (<, <=, etc) using 3-way comparison everywhere would likely solve the issue, but only if this is indeed the kind of issue.

**Comment by komiamiko** [comment]
As of [commit 2def959f7f32c8d78e2e4be7dd44370c2117b185](https://github.com/komiamiko/doodads/commit/2def959f7f32c8d78e2e4be7dd44370c2117b185), the performance has improved greatly. The change was fully implementing the 3-way comparator, and this seems to have fixed the exponential runtime issue.

# Unit test timings

While not a full benchmark, timing the unit tests should give a good idea of just how much improvement there was. The exact architecture I was using isn't very important, just look at the great improvement.

After the update:
* Full test battery: *15.533 ± 1.466 s* (5 samples)
* Only the fundamental sequence test: *0.367 ± 0.015 s* (5 samples)
* Only the fundamental sequence test, nerfed from N=19 to N=10: *0.340 ± 0.014 s* (5 samples)

Before the update:
* Full test battery: (no tests done)
* Only the fundamental sequence test: (no tests done)
* Only the fundamental sequence test, nerfed from N=19 to N=10: *34.820 ± 3.087 s* (5 samples)

# Unit test timings, again

So I actually left this comment unposted as I went and made what I thought was a small change (or a few changes), and it ended up significantly improving the performance once again. It's only a constant factor change, but I welcome the improvement. You can see it at [commit 371a3feeed0e7fa1ebec31acb6bde003bdfe7091](https://github.com/komiamiko/doodads/commit/371a3feeed0e7fa1ebec31acb6bde003bdfe7091)

* Full test battery†: *8.264 ± 0.584 s* (5 samples)
* Only the fundamental sequence test: *0.2224 ± 0.0037 s* (5 samples)

† Includes a new test. Be mindful of this when comparing with the previous measurements. Even with the new test, it's a large improvement.

# Conclusion

Performance is currently good. I'm closing the issue now, since there isn't a pressing need to optimize further.
